src/cache/cacher.rs

```diff
@@ -1,29 +1,52 @@
 //! This module provides the functionality to cache the aggregated results fetched and aggregated
 //! from the upstream search engines in a json format.
 
+use error_stack::Report;
+use futures::future::try_join_all;
 use md5::compute;
-use redis::{Client, Commands, Connection};
+use redis::{aio::ConnectionManager, AsyncCommands, Client, RedisError};
+
+use super::error::PoolError;
 
 /// A named struct which stores the redis Connection url address to which the client will
 /// connect to.
 ///
 /// # Fields
 ///
-/// * `redis_connection_url` - It stores the redis Connection url address.
+/// * `connection_pool` - It stores a pool of connections ready to be used.
+/// * `pool_size` - It stores the size of the connection pool (in other words the number of
+/// connections that should be stored in the pool).
+/// * `current_connection` - It stores the index of which connection is being used at the moment.
 pub struct RedisCache {
-    connection: Connection,
+    connection_pool: Vec<ConnectionManager>,
+    pool_size: u8,
+    current_connection: u8,
 }
 
 impl RedisCache {
     /// Constructs a new `SearchResult` with the given arguments needed for the struct.
     ///
     /// # Arguments
     ///
-    /// * `redis_connection_url` - It stores the redis Connection url address.
-    pub fn new(redis_connection_url: String) -> Result<Self, Box<dyn std::error::Error>> {
+    /// * `redis_connection_url` - It takes the redis Connection url address.
+    /// * `pool_size` - It takes the size of the connection pool (in other words the number of
+    /// connections that should be stored in the pool).
+    pub async fn new(
+        redis_connection_url: &str,
+        pool_size: u8,
+    ) -> Result<Self, Box<dyn std::error::Error>> {
         let client = Client::open(redis_connection_url)?;
-        let connection = client.get_connection()?;
-        let redis_cache = RedisCache { connection };
+        let mut tasks: Vec<_> = Vec::new();
+
+        for _ in 0..pool_size {
+            tasks.push(client.get_tokio_connection_manager());
+        }
+
+        let redis_cache = RedisCache {
+            connection_pool: try_join_all(tasks).await?,
+            pool_size,
+            current_connection: Default::default(),
+        };
```

Comment:
Why do we opt to manually create a connection pool? Usually, we would employ a connection pool library such as [`bb8`](https://docs.rs/bb8-redis/latest/bb8_redis/) or [`r2d2`](https://github.com/sorccu/r2d2-redis) to manage connection pooling, which tends to be a more common approach.

(To be honest, I haven't had much experience with Redis, so all of the opinions are based on some internet research:))

Reply:
Yes, I totally agree with you, I would also prefer the same, but it seems that none of the pooling libraries like redis-rs, bb8 or r2d2 provide all the 3 things together which is asynchronous connections, connection-manager and pooling. That's why I had to choose this method and had to manage pooling on our own. Also, I want the 3 features together because I want to provide a very low latency and performant system, that's why I am wanting to have the 3 things together. :slightly_smiling_face:

Reply:
I don't think this implementation of the connection pool is correct. The current implementation doesn't seem to be thread-safe. If multiple threads call `cached_json` simultaneously, they might interfere with each other's usage of the `current_connection` index and the `connection_pool`.

In a typical connection pool, connections are reused as much as possible to avoid the overhead of establishing a new connection. However, in the code provided at https://github.com/neon-mmd/websurfx/blob/a799af312525cb198fd744826ca4118f26bef3c8/src/server/routes.rs#L162, we create a new `RedisCache` for each query.

> Also, I want the 3 features together because I want to provide a very low latency and performant system, that's why I am wanting to have the 3 things together

I don't think the current implementation is faster than `r2d2` or `bb8`. Furthermore, I don't think our system would benefit significantly from optimizing this aspect. Therefore, I suggest conducting benchmarking before prioritizing optimization efforts

Reply:
Yes, you are right once again :joy: Actually I have worked on this issue by fixing it in the next part of the series of PRs, so I would also suggest taking a look at it as well :slightly_smiling_face: .